### PR TITLE
fix(healthcheck) retry locking protected fn

### PR DESCRIPTION
**Pull request by locao**
When ngx.sleep API is not available (e.g. in the log phase) it's not possible to lock using lua-resty-lock and functions that must run protected were failing.

This change adds a retry method that starts a new light thread that has access to ngx.sleep and will succeed to lock.

Fix Kong/kong#5137

### CODE REVIEW

File: lib/resty/healthcheck.lua [hunk 1]
```diff
@@ -436,20 +432,47 @@ local function locking_target(self, ip, port, hostname, fn)
...
     self:log(ERR, "failed to release lock '", lock_key, "': ", err)
   end
 
   return final_ok, final_err
+
+end
+
+
+-- Run the given function holding a lock on the target.
+-- WARNING: the callback will run unprotected, so it should never
```
Comment (kikito) [inline lib/resty/healthcheck.lua]: I don't understand why this comment is needed. What's the advantage of invoking the function in an unprotected way? I'd rather use `pcall` and remove this comment

**Comment by hishamhm** [comment]
@locao I believe `locking_target_list` needs the same thing.

File: lib/resty/healthcheck.lua (hunk 1 above)
Comment (locao) [reply lib/resty/healthcheck.lua]: Good call! Fixed, thanks.

**Comment by dabue** [comment]
The timer delays adding targets, so it will get an error if use the get_target_status function to get the status of a node before the targets assigned.
It's the best to use the get_target_status function by pcall, and use the default is_healthy status as result.

**Comment by dabue** [comment]
the locking_target_list delays the initialization and update of the targets.

**Comment by Tieske** [comment]
@dabue please do not respond to closed issues/PR's please open a new one if needed

**Pull request merged**